src/task_impl/atomic_task.rs

```diff
@@ -86,90 +73,96 @@ impl AtomicTask {
+
+                    // Release the lock. If the state transitioned to include
+                    // the `NOTIFYING` bit, this means that a notify has been
+                    // called concurrently, so we have to remove the task and
+                    // notify it.`
+                    //
+                    // Start by assuming that the state is `REGISTERING` as this
+                    // is what we jut set it to.
+                    let mut curr = REGISTERING;
+
+                    // If a task has to be notified, it will be set here.
+                    let mut notify = None;
+
+                    loop {
+                        let res = self.state.compare_exchange(
+                            curr, WAITING, AcqRel, Acquire);
+
+                        match res {
+                            Ok(_) => {
+                                // The atomic exchange was successful, now break
+                                // out of the loop os that task stored in
+                                // `notify` can be notified (if set).
+                                break;
+                            }
+                            Err(actual) => {
+                                // Update `curr` for the next iteration of the
+                                // loop
+                                curr = actual;
+                            }
+                        }
+
+                        // Since we aren't using the weak variant of the atomic
+                        // operation, the only possible option for `curr` is to
+                        // also include the `NOTIFYING` bit.
+                        debug_assert_eq!(curr, curr | NOTIFYING);
+
+                        // Take the task to notify once the atomic operation has
+                        // completed.
+                        notify = (*self.task.get()).take();
+                    }
+
+                    if let Some(task) = notify {
+                        task.notify();
                     }
                 }
             }
-            LOCKED_WRITE | LOCKED_WRITE_NOTIFIED => {
-                // A thread is concurrently calling `register`. This shouldn't
-                // happen as it doesn't really make much sense, but it isn't
-                // unsafe per se. Since two threads are concurrently trying to
-                // update the task, it's undefined which one "wins" (no ordering
-                // guarantees), so we can just do nothing.
-            }
-            state => {
-                debug_assert!(state != LOCKED_WRITE, "unexpected state LOCKED_WRITE");
-                debug_assert!(state != LOCKED_WRITE_NOTIFIED, "unexpected state LOCKED_WRITE_NOTIFIED");
-
-                // Currently in a read locked state, this implies that `notify`
-                // is currently being called on the old task handle. So, we call
-                // notify on the new task handle
+            NOTIFYING => {
```

Comment:
Shouldn't it also call `task.notify()` if we read `REGISTERING | NOTIFYING`?

Reply:
I'm treating `REGISTERING` being set when calling `register` as a caller bug. If you want, I can add a `panic` in this case?